integration/release_upgrade_linux_test.go

```diff
@@ -0,0 +1,364 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	criruntime "k8s.io/cri-api/pkg/apis/runtime/v1"
+
+	cri "github.com/containerd/containerd/integration/cri-api/pkg/apis"
+	"github.com/containerd/containerd/integration/images"
+	"github.com/containerd/containerd/integration/remote"
+)
+
+// upgradeVerifyCaseFunc is used to verify the behavior after upgrade.
+type upgradeVerifyCaseFunc func(t *testing.T, criRuntimeService cri.RuntimeService)
+
+// TODO: Support Windows
+func TestUpgrade(t *testing.T) {
+	previousReleaseBinDir := t.TempDir()
+	downloadPreviousReleaseBinary(t, previousReleaseBinDir)
+
+	t.Run("recover", runUpgradeTestCase(previousReleaseBinDir, shouldRecoverAllThePodsAfterUpgrade))
+	// TODO:
+	// Add exec/stats/stop-existing-running-pods/...
+}
+
+func runUpgradeTestCase(
+	previousReleaseBinDir string,
+	setupUpgradeVerifyCase func(t *testing.T, criRuntimeService cri.RuntimeService, criImageService cri.ImageManagerService) upgradeVerifyCaseFunc,
+) func(t *testing.T) {
+	return func(t *testing.T) {
+		workDir := t.TempDir()
+
+		t.Log("Install config for previous release")
+		previousReleaseCtrdConfig(t, previousReleaseBinDir, workDir)
+
+		t.Log("Starting the previous release's containerd")
+		previousCtrdBinPath := filepath.Join(previousReleaseBinDir, "bin", "containerd")
+		previousProc := newCtrdProc(t, previousCtrdBinPath, workDir)
+
+		ctrdLogPath := previousProc.logPath()
+		t.Cleanup(func() {
+			dumpFileContent(t, ctrdLogPath)
+		})
+
+		require.NoError(t, previousProc.isReady())
+
+		needToCleanup := true
+		t.Cleanup(func() {
+			if t.Failed() && needToCleanup {
+				t.Logf("Try to cleanup leaky pods")
+				cleanupPods(t, previousProc.criRuntimeService(t))
+			}
+		})
+
+		t.Log("Prepare pods for current release")
+		upgradeCaseFunc := setupUpgradeVerifyCase(t, previousProc.criRuntimeService(t), previousProc.criImageService(t))
+		needToCleanup = false
+
+		t.Log("Gracefully stop previous release's containerd process")
+		require.NoError(t, previousProc.kill(syscall.SIGTERM))
+		require.NoError(t, previousProc.wait())
```

Comment:
Maybe a timeout here?  Or (see other comment about pdeathsig) have the parent goroutine in a locked OS thread exit, causing SIGKILL to be sent?

Reply:
I changed the `wait()` to `wait(timeout time.Duration)`. And also add new goroutine to wait for containerd's exit event with locked OS thread. PTAL